src/element/mod.rs

```diff
@@ -256,13 +256,16 @@ impl IonEq for Element {
 /// An `(annotations, value)` pair representing an Ion value.
 #[derive(Debug, Clone)]
 pub struct Element {
-    annotations: Vec<Symbol>,
+    annotations: Annotations,
```

Comment:
(Question) Should we rather have `Annotation` type defined which will be a `Symbol` and then we can return a `Vec<Annotation>` for `Element::annotations()` and one can perform all the `Vec` related operations on it?
I guess my question is which part are we trying to make abstract, do we want to make the entire `Vec` opaque or just provide something like `Annotation`?

Reply:
The goal is to abstract away the `Vec` itself. On the stack, a `Vec` is 24 bytes on a 64-bit machine:
* 8 bytes for its current length (`usize`)
* 8 bytes for its capacity (`usize`)
* 8 bytes for a pointer to its array on the heap.

An `Element` is a `(Value, Annotations)` pair, which means that its stack size is 24 bytes + the size of a `Value`. Trimming down `Annotations` would also trim down `Element`, making it faster to construct/move. (See the [_Type Sizes_ section](https://nnethercote.github.io/perf-book/type-sizes.html) of the Rust performance book for more details.)

Most Ion values don't have any annotations; I believe that replacing `Vec` with something like `ThinVec` would be beneficial. Of course, we would need to write some benchmarks to be sure. For now, I'm just wrapping `Vec` in an opaque type so we can swap out its implementation later without causing problems.

> Should we rather have Annotation type defined which will be a Symbol

Having an `Annotation` type is an interesting idea, though I imagine its meaning will always be very closely aligned with the `Symbol` type. (`Symbol` is a custom type of ours anyway, so we can already modify its implementation if needed.)